ikarus/controlRoutines/adaptiveStepSizing.hh

```diff
@@ -0,0 +1,90 @@
+#include <Eigen/Dense>
+#include <Eigen/Sparse>
+
+#include <ikarus/linearAlgebra/nonLinearOperator.hh>
+#include <ikarus/solver/nonLinearSolver/solverInfos.hh>
+#include <ikarus/utils/fancyActivationFunctions.hh>
+
+namespace Ikarus {
+
+  template <typename ActivationFunction = Ikarus::AdaptiveStepSizing::LinearPiecewiseFunction,
+            typename ScalarType         = double>
+  struct DefaultAdaptiveStepSizing {
+    template <typename NonLinearOperator>
+    void operator()(const NonLinearSolverInformation& solverInfo, const NonLinearOperator&,
+                    SubsidiaryArgs& subsidiaryArgs, const ActivationFunction&) {
+      if (targetIterations != 0 && subsidiaryArgs.actStep > 0) {
+        ScalarType actIte       = solverInfo.iterations;
+        subsidiaryArgs.stepSize = sqrt(static_cast<ScalarType>(targetIterations) / actIte) * subsidiaryArgs.stepSize;
+      }
+    }
+    int targetIterations{0};
+  };
+
+  template <typename ActivationFunction = Ikarus::AdaptiveStepSizing::LinearPiecewiseFunction,
+            typename ScalarType         = double>
+  struct FancyAdaptiveStepSizing {
+    template <typename NonLinearOperator>
+    void operator()(const NonLinearSolverInformation& solverInfo, NonLinearOperator& nonLinearOperator,
+                    SubsidiaryArgs& subsidiaryArgs, const ActivationFunction& activationFunction) const {
+      const auto& K = nonLinearOperator.derivative();
+      auto act_step = subsidiaryArgs.actStep;
+
+      auto K_size = K.rows();
+
+      /// Construct matrix operation object using the wrapper class DenseSymMatProd
+      Spectra::DenseSymMatProd<double> op(K);
+
+      /// Construct eigen solver object, requesting the largest three eigenvalues
+      auto num_eigs  = 3;  // TODO: Generalize it
+      auto eigs_conv = 2 * num_eigs;
+      Spectra::SymEigsSolver<Spectra::DenseSymMatProd<double>> eigs(op, num_eigs, eigs_conv);
+
+      /// Initialize and compute
+      eigs.init();
+      // eigs.compute(Spectra::SortRule::SmallestAlge);
+      eigs.compute(Spectra::SortRule::LargestAlge);  // TODO: Change it to SmallestAlge
+      Eigen::VectorXd evalues(num_eigs);
+      if (eigs.info() == Spectra::CompInfo::Successful) evalues = eigs.eigenvalues();
+
+      std::cout << "Eigenvalues found:\n" << evalues << std::endl;
+
+      std::vector<double> evstd;
+      for (auto i = 0U; i < evalues.size(); ++i)
+        if (std::abs(evalues[i] - 1.0)
+            < 1e-8)  // TODO: Fix it such that it works for both reduced and full matrices correctly
+          evstd.push_back(evalues[i] * 1e10);
+        else
+          evstd.push_back(evalues[i]);
+
+      std::sort(evstd.begin(), evstd.end(), std::less<double>());
+      for (auto i = 0U; i < evalues.size(); ++i)
+        evalues[i] = evstd[i];
+
+      std::cout << "Eigenvalues sorted:\n" << evalues << std::endl;
+
+      /// store eigenvalues of each increment in a matrix
+      subsidiaryArgs.eigenValues.col(act_step) = evalues;

```

Comment:
maybe create a small class here? for the eigenvalues of each step. see also line 84. i is a bit ugly to always access the current eigenvalues like that. creating a wrapper cleans this maybe